rust/exchange_rate/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+use std::cell::RefCell;
+use std::collections::HashMap;
+
+type Timestamp = u64;
+type Rate = f32;
+
+#[derive(CandidType, Clone, Deserialize, Debug, Eq, Hash, PartialEq, Serialize)]
+pub struct HttpHeader {
+    pub name: String,
+    pub value: String,
+}
+
+#[derive(Clone, Debug, PartialEq, CandidType, Eq, Hash, Serialize, Deserialize)]
+pub enum HttpMethod {
+    GET,
+}
+
+#[derive(CandidType, Deserialize, Debug)]
+pub struct CanisterHttpRequestArgs {
+    pub url: String,
+    pub headers: Vec<HttpHeader>,
+    pub body: Option<Vec<u8>>,
+    pub http_method: HttpMethod,
+    pub transform_method_name: Option<String>,
+}
+
+#[derive(CandidType, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct CanisterHttpResponsePayload {
+    pub status: u64,
+    pub headers: Vec<HttpHeader>,
+    pub body: Vec<u8>,
+}
+
+thread_local! {
+    pub static FETCHED: RefCell<HashMap<Timestamp, Rate>>  = RefCell::new(HashMap::new());
+    pub static REQUESTED: RefCell<Queue<Timestamp>> = RefCell::new(Queue::new());
+
+    pub static RESPONSE_HEADERS_SANTIZATION: Vec<&'static str> = vec![
+        "x-mbx-uuid",
+        "x-mbx-used-weight",
+        "x-mbx-used-weight-1m",
+        "Date",
+        "Via",
+        "X-Amz-Cf-Id",
+    ];
+}
+
+/// Canister heartbeat. Process one item in queue
+// #[export_name = "canister_heartbeat"]
+#[heartbeat]
+async fn heartbeat() {
+    ic_cdk::println!("Heartbeat method invoked.");
+    get_next_rate().await;
+}
+
+/*
+Get rates for a time range defined by start time and end time. This function can be invoked
+as HTTP update call.
+*/
+#[update]
+#[candid_method(update)]
+async fn get_rates(start: Timestamp, end: Timestamp) -> HashMap<Timestamp, Rate> {
+    // round down start time and end time to the minute (chop off seconds), to be checked in the hashmap
+    let start_min = start / 60;
+    let end_min = end / 60;
+
+    // compose a return structure
+    let mut fetched = HashMap::new();
+
+    // pull available ranges from hashmap
+    FETCHED.with(|map_lock| {
+        let map = map_lock.borrow();
+        for requested_min in start_min..end_min {
+            let requested = requested_min * 60;
+            if map.contains_key(&requested) {
+                // The fetched slot is within user requested range. Add to result for later returning.
+                fetched.insert(requested, map.get(&requested).unwrap().clone());
+            } else {
+                // asynchoronously request downloads for unavailable ranges
+
+                // Simply putting the request to request queue. This queue will likely
+                // have duplicate entries, if users request same range of data multiple times.
+                // Double downloading is avoided right before the time of downloading by checking
+                // whether the data already exists in FETCHED map.
+                REQUESTED.with(|requested_lock| {
+                    let mut queue = requested_lock.borrow_mut();
```

Comment:
Rust idiom is to reuse names, so this could be "let mut queue  = queue.borrow_mut();"